internal/translation/datafederation/conversion_endpoints.go

```diff
@@ -58,3 +65,39 @@ func endpointToAtlas(ep *DatafederationPrivateEndpointEntry) *admin.PrivateNetwo
 		Type:       pointer.MakePtrOrNil(ep.Type),
 	}
 }
+
+type DataFederationPrivateEndpoint struct {
+	AKO, Atlas, LastApplied *DatafederationPrivateEndpointEntry
+}
+
+func MapDatafederationPrivateEndpoints(projectID string, df *akov2.AtlasDataFederation, atlasEndpoints []*DatafederationPrivateEndpointEntry) (map[string]*DataFederationPrivateEndpoint, error) {
+	var lastApplied akov2.AtlasDataFederation
+	if ann, ok := df.GetAnnotations()[customresource.AnnotationLastAppliedConfiguration]; ok {
+		err := json.Unmarshal([]byte(ann), &lastApplied.Spec)
+		if err != nil {
+			return nil, fmt.Errorf("error reading data federation from last applied annotation: %w", err)
+		}
+	}
```

Comment:
nit

Being purist I would do this outside of the translation layer. Basically the translation layer should be Kubernetes unaware. We could pass this as an argument.

Not blockling the PR due to this, but design wise I think is better to not tangle the translation with the concept of annotations in kube objects.

Reply:
I actually don't agree in this case.

This is definitely part of the translation layer, it introspects public types and its internal state, in this case annotation, and generates internal types to be consumed by the controller, in this case a map.

Reply:
> the translation layer should be Kubernetes unaware. 

I agree but this is not the case here. We are introspecting the structure of the public type. The annotation is a simple map, we encode previous state in here using JSON, there is nothing Kubernetes specific in here.

Reply:
Actually, you are right, we are not using any Kubernetes interfaces or APIs here. Annotations are just maps, so yes, I guess is fine.